test/unit/s3gateway_test.js

```diff
@@ -265,7 +267,100 @@ function testEscapeURIPathPreservesDoubleSlashes() {
+                    AccessKeyId: 'AN_ACCESS_KEY_ID',
+                    Expiration: '2017-05-17T15:09:54Z',
+                    RoleArn: 'TASK_ROLE_ARN',
+                    SecretAccessKey: 'A_SECRET_ACCESS_KEY',
+                    Token: 'A_SECURITY_TOKEN',
+                });
+            }
+        });
+    };
+    var r = {
+        "headersOut" : {
+            "Accept-Ranges": "bytes",
+            "Content-Length": 42,
+            "Content-Security-Policy": "block-all-mixed-content",
+            "Content-Type": "text/plain",
+            "X-Amz-Bucket-Region": "us-east-1",
+            "X-Amz-Request-Id": "166539E18A46500A",
+            "X-Xss-Protection": "1; mode=block"
+        }
+    };
+    r.log = function(msg) {
+        console.log(msg);
+    }
+
+    await s3gateway.fetchCredentials(r);
+
+    if (globalThis.recordedUrl !== 'http://localhost/example') {
+        throw 'No or wrong ECS credentials fetch URL recorded: ' + globalThis.recordedUrl;
+    }
+}
+
+async function testEc2CredentialRetrieval() {
+    process.env['S3_ACCESS_KEY_ID'] = undefined;
+    process.env['AWS_CONTAINER_CREDENTIALS_ABSOLUTE_URI'] = undefined;
+    globalThis.ngx.fetch = function (url, options) {
+        if (url === 'http://169.254.169.254/latest/api/token' && options && options.method === 'PUT') {
+            return Promise.resolve({
+                text: function () {
+                    return Promise.resolve('A_TOKEN');
+                },
+            });
+        } else if (url === 'http://169.254.169.254/latest/meta-data/iam/security-credentials/s3access') {
+            if (options && options.headers && options.headers['x-aws-ec2-metadata-token'] === 'A_TOKEN') {
+                return Promise.resolve({
+                    json: function () {
+                        globalThis.credentialsIssues = true;
+                        return Promise.resolve({
+                            AccessKeyId: 'AN_ACCESS_KEY_ID',
+                            Expiration: '2017-05-17T15:09:54Z',
+                            RoleArn: 'TASK_ROLE_ARN',
+                            SecretAccessKey: 'A_SECRET_ACCESS_KEY',
+                            Token: 'A_SECURITY_TOKEN',
+                        });
+                    },
+                });
+            } else {
+                throw 'Invalid token passed: ' + options.headers['x-aws-ec2-metadata-token'];
+            }
+        } else {
+            throw 'Invalid request URL: ' + url;
+        }
+    };
+    var r = {
+        "headersOut" : {
+            "Accept-Ranges": "bytes",
+            "Content-Length": 42,
+            "Content-Security-Policy": "block-all-mixed-content",
+            "Content-Type": "text/plain",
+            "X-Amz-Bucket-Region": "us-east-1",
+            "X-Amz-Request-Id": "166539E18A46500A",
+            "X-Xss-Protection": "1; mode=block"
+        }
+    };
+    r.log = function(msg) {
+        console.log(msg);
+    }
+
```

Comment:
Here we will need to add an additional function to the `r` instance when running as a unit test in order to avoid an exception (which is what is causing the segfault - a patch will be included in the next njs release):
```javascript
    r.return = function(code, body) {
        console.log(`RESPONSE RETURN: ${code} ${body}`);
    }
```

Reply:
I will add the function, but I don't think it is valuable to log the whole response, isn't it? I might simply add an "assertion" that the correct code is returned, ok? :)

Reply:
Great! That dealt with one of the two segfaults that were happening. Using an assertion here makes a lot of sense.